components/webapps/blacklist-whitelist-service/src/main/java/com/wso2telco/services/bw/Queries.java

```diff
@@ -56,360 +51,262 @@
-
-		LOG.debug("provisionWhiteListedNumber Triggerd  jsonBody :" + jsonBody + " , msisdn :" + msisdn);
-
-		Gson gson = new GsonBuilder().serializeNulls().create();
-
-		WhiteList whiteListReq = gson.fromJson(jsonBody, WhiteList.class);
-
-		String subscriptionID = whiteListReq.getSubscriptionID();
-		String apiID = whiteListReq.getAPIID();
-		String applicationID = whiteListReq.getApplicationID();
-		final StringBuilder errorMSG = new StringBuilder();
-
-		errorMSG.append("{").append("\"Failed\":".intern()).append("{").append("\"messageId\":\"".intern())
-				.append("Whitelisted Number".intern()).append("\",");
-		errorMSG.append("\"text\":\"").append("Whitelisted Number could not be added to the system ".intern())
-				.append("\",");
-		errorMSG.append("\"variables\":\"").append(msisdn).append("\"").append("}}");
-
-		if (subscriptionID != null && apiID != null && applicationID != null) {
+	@Path("/bulkAdd/{app}/{api}")
+	@Consumes(MediaType.MULTIPART_FORM_DATA)
+	public Response bulkBlacklist(@PathParam("app") String app, @PathParam("api") String api, @Multipart("file") InputStream uploadedInputStream, @Multipart("action") String action, @Multipart("sp") String sp, @Multipart("user") String user) {
 
-			try {
-
-				WhiteListDTO whiteListDTO = new WhiteListDTO();
-				whiteListDTO.setApiID(apiID);
-				whiteListDTO.setApplicationID(applicationID);
-				//whiteListDTO.setSubscriptionID(subscriptionID);
-				whiteListDTO.setUserMSISDN(msisdn);
+		APIBlacklistWhitelistResponseDTO response;
+		BlacklistWhitelistDTO blacklistWhitelistDTO = new BlacklistWhitelistDTO();
+		Gson gson = new Gson();
+		List<String> msisdnList = new ArrayList<>();
+		BufferedReader br;
+		int limit = 5000; //max number of records to process
 
-				blackListWhiteListService.whiteListSubscription(whiteListDTO);
+		app = replaceAllWithWildcard(app);
+		api = replaceAllWithWildcard(api);
+		sp = replaceAllWithWildcard(sp);
 
-				final StringBuilder succMSG = new StringBuilder();
+		blacklistWhitelistDTO.setAppId(app);
+		blacklistWhitelistDTO.setApiID(api);
+		blacklistWhitelistDTO.setAction(action);
+		blacklistWhitelistDTO.setServiceProvider(sp);
+		blacklistWhitelistDTO.setUser(user);
 
-				succMSG.append("{").append("\"Success\":").append("{").append("\"messageId\":\"")
-						.append("WhiteListed Number".intern()).append("\",");
-				succMSG.append("\"text\":\"").append("WhiteListed Number Successfully Added to the system ".intern())
-						.append("\",");
-				succMSG.append("\"variables\":\"").append(msisdn).append("\"").append("}}");
 
-				return Response.status(Response.Status.OK).entity(succMSG.toString()).build();
+		br = new BufferedReader(new InputStreamReader(uploadedInputStream, StandardCharsets.UTF_8));
 
-			}catch( InvalidMSISDNException e1){
-				return Response.status(Response.Status.BAD_REQUEST).entity(e1.getErrorType()).build();
+		try {
+			while (br.ready() && limit > 0) {
+				msisdnList.add(br.readLine());
```

Comment:
Use the msisdn validations here within the while loop. No point of iterating the same list again the in com.wso2telco.dep.operatorservice.service.BlackListWhiteListService#blacklist to find valid and invalid msisdns. It is a performance hit

Reply:
it was decided to have all business logic in service level